[JSC] DFG GetGlobalLexicalVar / GetGlobalVar 's emptiness should be monotonic

Constellation opened the pull request:
#### 62a0d0bc837e521bbbd3fd1067a0d7875f3d42c7
<pre>
[JSC] DFG GetGlobalLexicalVar / GetGlobalVar &apos;s emptiness should be monotonic
<a href="https://bugs.webkit.org/show_bug.cgi?id=286246">https://bugs.webkit.org/show_bug.cgi?id=286246</a>
<a href="https://rdar.apple.com/143243411">rdar://143243411</a>

Reviewed by Yijia Huang.

They are tied to JSGlobalObject. And its slot&apos;s emptiness should be
monotonically changing. Thus if we found it has non empty value, we can
say it does not have empty value.

* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):

Canonical link: <a href="https://commits.webkit.org/289680@main">https://commits.webkit.org/289680@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7ad1808366f2169ca4a8d0ada0ba7572719859e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87646 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92511 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38390 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7542 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15332 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67674 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25417 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90648 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5755 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79308 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48038 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5543 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33709 "Found 60 new test failures: editing/undo/redo-reapply-edit-command-crash.html fast/dom/Window/post-message-large-array-buffer-should-not-crash.html fast/repaint/emoji-glyph-overflow-repaint-fail.html fonts/monospace.html fonts/sans-serif.html fonts/serif.html fonts/unicode-character-font-crash.html fonts/use-typo-metrics-1.html http/tests/security/clipboard/copy-paste-html-cross-origin-iframe-across-origin.html http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-script-in-iframe.html ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37503 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80443 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75943 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34573 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94397 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86421 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14814 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10860 "Found 4 new test failures: fast/canvas/check-stale-putImageData.html ipc/create-media-source-with-invalid-constraints-crash.html js/exception-with-handler-inside-eval-with-dynamic-scope.html svg/custom/use-image-in-g.svg (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76524 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15068 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75164 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75750 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20119 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18555 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7765 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13669 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14830 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20131 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108914 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14574 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26191 "Found 17 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/function-apply-many-args.js.layout-no-llint, stress/regexp-escape-oom.js.default, wasm.yaml/wasm/gc/array_new_data.js.default-wasm, wasm.yaml/wasm/gc/array_new_data.js.wasm-bbq, wasm.yaml/wasm/gc/array_new_data.js.wasm-collect-continuously, wasm.yaml/wasm/gc/array_new_data.js.wasm-eager, wasm.yaml/wasm/gc/array_new_data.js.wasm-eager-jettison, wasm.yaml/wasm/gc/array_new_data.js.wasm-no-cjit, wasm.yaml/wasm/gc/array_new_data.js.wasm-slow-memory, wasm.yaml/wasm/stress/array-element-creation.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18018 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16356 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -5042,16 +5042,36 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+            // So long as it is not unlinked.
+            if (!m_graph.m_plan.isUnlinked()) {
+                JSValue concurrentlyRead = node->variablePointer()->get();
+                if (concurrentlyRead != JSValue()) {
```
nit: `!concurrentlyRead.isEmpty()`

Constellation replied:
Changed it to boolean check `if (JSValue oncurrentlyRead = node->variablePointer()->get())` as it returns false for jsempty (and it is the pattern we are using in JSC).

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -5042,16 +5042,36 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+            // So long as it is not unlinked.
+            if (!m_graph.m_plan.isUnlinked()) {
+                JSValue concurrentlyRead = node->variablePointer()->get();
+                if (concurrentlyRead != JSValue()) {
```
Ditto

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] DFG GetGlobalLexicalVar / GetGlobalVar 's emptiness should be monotonic
https://bugs.webkit.org/show_bug.cgi?id=286246
rdar://143243411

Reviewed by Yijia Huang.

They are tied to JSGlobalObject. And its slot's emptiness should be
monotonically changing. Thus if we found it has non empty value, we can
say it does not have empty value.

* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):

Canonical link: https://commits.webkit.org/289680@main
